### PR TITLE
Use arrays for line items order population

### PR DESCRIPTION
**Pull request by BenMorganIO**

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://benmorganio-use-arrays-for-line-items-order-population.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://benmorganio-use-arrays-for-line-items-order-population.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: core/lib/spree/core/importer/order.rb [hunk 1]
```diff
@@ -97,21 +97,66 @@ def self.create_shipments_from_params(shipments_hash, order)
...
+                extra_params = line_items[k].except(:variant_id, :quantity, :sku)
+                line_item = ensure_variant_id_from_params(line_items[k])
+                line_item = order.contents.add(Spree::Variant.find(line_item[:variant_id]), line_item[:quantity])
+                # Raise any errors with saving to prevent import succeeding with line items failing silently.
+                if extra_params.present?
+                  line_item.update_attributes!(extra_params)
+                else
+                  line_item.save!
+                end
+              rescue Exception => e
```
Comment (BenMorganIO) [inline core/lib/spree/core/importer/order.rb]: @houndci That is so staying.

File: core/lib/spree/core/importer/order.rb [hunk 2]
```diff
@@ -97,21 +97,66 @@ def self.create_shipments_from_params(shipments_hash, order)
...
-            begin
-              extra_params = line_items_hash[k].except(:variant_id, :quantity, :sku)
-              line_item = ensure_variant_id_from_params(line_items_hash[k])
-              line_item = order.contents.add(Spree::Variant.find(line_item[:variant_id]), line_item[:quantity])
-              # Raise any errors with saving to prevent import succeeding with line items failing silently.
-              if extra_params.present?
-                line_item.update_attributes!(extra_params)
-              else
-                line_item.save!
+        def self.create_line_items_from_params(line_items, order)
```
Comment (BenMorganIO) [inline core/lib/spree/core/importer/order.rb]: Keep the perceived complexity. It'll be way easier to delete the code in the future.

**Comment by BenMorganIO** [comment]
I thought about making the deprecation warnings small and compact, but wow, when there's tons of whitespace, those deprecation warnings are so easy to read.

File: core/lib/spree/core/importer/order.rb [hunk 3]
```diff
@@ -97,21 +97,71 @@ def self.create_shipments_from_params(shipments_hash, order)
...
+                variant = Spree::Variant.find(line_item[:variant_id])
+                line_item = order.contents.add(variant, line_item[:quantity])
+                # Raise any errors with saving to prevent import succeeding with line items
+                # failing silently.
+                if extra_params.present?
+                  line_item.update_attributes!(extra_params)
+                else
+                  line_item.save!
+                end
+              rescue Exception => e
```
Comment (BenMorganIO) [inline core/lib/spree/core/importer/order.rb]: Yes @houndci, you've barked up this tree before...

File: core/lib/spree/core/importer/order.rb (hunk 1 above)
Comment (BenMorganIO) [inline core/lib/spree/core/importer/order.rb]: @houndci bite me.

File: core/lib/spree/core/importer/order.rb (hunk 3 above)
Comment (BenMorganIO) [inline core/lib/spree/core/importer/order.rb]: @houndci woof.

File: core/lib/spree/core/importer/order.rb (hunk 1 above)
Comment (Senjai) [inline core/lib/spree/core/importer/order.rb]: @BenMorganIO You really should never rescue Exception. There are some cases, but this isn't one of them. Especially when you're doing a lot of stuff in this method and so much can go wrong.

**Comment by JDutil** [comment]
:+1:

**Comment by BenMorganIO** [comment]
Specs broke on updating the guides... No idea why.

**Comment by JDutil** [comment]
Thats a known frail spec still having figure out a fix yet

**Comment by BenMorganIO** [comment]
Awesome. She's mergeable :D

**Comment by JDutil** [comment]
Could you add this the 3.0 release notes as well:
https://github.com/spree/spree/blob/master/guides/content/release_notes/3_0_0.md

File: core/lib/spree/core/importer/order.rb [hunk 4]
```diff
@@ -97,21 +97,71 @@ def self.create_shipments_from_params(shipments_hash, order)
...
-                line_item.update_attributes!(extra_params)
-              else
-                line_item.save!
+        def self.create_line_items_from_params(line_items, order)
+          return {} unless line_items
+          case line_items
+          when Hash
+            ActiveSupport::Deprecation.warn(<<-EOS, caller)
+              Passing a hash is now deprecated, it is recommended that you pass
+              it as an array instead.
```
Comment (JDutil) [inline core/lib/spree/core/importer/order.rb]: Can you indicate this will be removed in 3.1

**Pull request closed without merging**